core/src/dird/dird.cc

```diff
@@ -301,11 +297,17 @@ int main(int argc, char* argv[])
 
   int pidfile_fd = -1;
 #if !defined(HAVE_WIN32)
-  if (!test_config && !foreground && pidfile_path) {
-    pidfile_fd = CreatePidFile("bareos-dir", pidfile_path);
+  if (!test_config && !foreground && !pidfile_path.empty()) {
+    pidfile_fd = CreatePidFile("bareos-dir", pidfile_path.c_str());
   }
 #endif
   // See if we want to drop privs.
+  char* uid = nullptr;
+  if (!user.empty()) { uid = user.data(); }
```

Comment:
Here, `str.data()` is used, while more down, `str.c_str()` is used. I think it would make sense to use the same function always, and as we do want to have a nullterminated c string, I would opt for using `c_str()` everywhere.

While nowadays both are identical, historically only the `c_str()` was guaranteed to have a \0 at the end.

Reply:
while I would agree about keeping it the same, the problem goes back to the type returned. 
Each situation demands a different type that can only be fulfilled by either one of the two functions. 
For this specific part, we need a `char*` to give to `uid` that can be only obtained with `.data()` while later in the code, we need a `const char*` that can only be obtained with `.c_str()`

Reply:
Ah, I see. That of course makes sense then. My checks for the difference between the two functions told me that  `The returned array is null-terminated, that is, data() and [c_str()](https://en.cppreference.com/w/cpp/string/basic_string/c_str) perform the same function.`